web/profiles/custom/yalesites_profile/modules/custom/ys_core/src/Plugin/Block/YaleSitesBreadcrumbBlock.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+namespace Drupal\ys_core\Plugin\Block;
+
+use Drupal\Core\Block\BlockBase;
+use Drupal\Core\Breadcrumb\BreadcrumbBuilderInterface;
+use Drupal\Core\Plugin\ContainerFactoryPluginInterface;
+use Drupal\Core\Routing\RouteMatchInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Provides a block to display the breadcrumbs.
+ *
+ * @Block(
+ *   id = "ys_breadcrumb_block",
+ *   admin_label = @Translation("YaleSites Breadcrumbs"),
+ *   category = @Translation("YaleSites Core"),
+ * )
+ */
+class YaleSitesBreadcrumbBlock extends BlockBase implements ContainerFactoryPluginInterface {
+
+  /**
+   * The breadcrumb manager.
+   *
+   * @var \Drupal\Core\Breadcrumb\BreadcrumbBuilderInterface
+   */
+  protected $breadcrumbManager;
+
+  /**
+   * The current route match.
+   *
+   * @var \Drupal\Core\Routing\RouteMatchInterface
+   */
+  protected $routeMatch;
+
+  /**
+   * Constructs a new SystemBreadcrumbBlock object.
+   *
+   * @param array $configuration
+   *   A configuration array containing information about the plugin instance.
+   * @param string $plugin_id
+   *   The plugin_id for the plugin instance.
+   * @param mixed $plugin_definition
+   *   The plugin implementation definition.
+   * @param \Drupal\Core\Breadcrumb\BreadcrumbBuilderInterface $breadcrumb_manager
+   *   The breadcrumb manager.
+   * @param \Drupal\Core\Routing\RouteMatchInterface $route_match
+   *   The current route match.
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, BreadcrumbBuilderInterface $breadcrumb_manager, RouteMatchInterface $route_match) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->breadcrumbManager = $breadcrumb_manager;
+    $this->routeMatch = $route_match;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('breadcrumb'),
+      $container->get('current_route_match')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function build() {
+    $breadcrumbs = $this->breadcrumbManager->build($this->routeMatch)->getLinks();
+    $links = [];
+
+    foreach ($breadcrumbs as $breadcrumb) {
+      array_push($links, [
+        'title' => $breadcrumb->getText(),
+        'url' => $breadcrumb->getUrl()->toString(),
+      ]);
+    }
+
+    return [
+      '#theme' => 'ys_breadcrumb_block',
+      '#items' => $links,
+    ];
+  }
+
+  /**
+   * Turns off block cache.
+   *
+   *   Returns 0 to disable block cache.
+   */
+  public function getCacheMaxAge() {
+    return 0;
+  }
```

Comment:
I'm not a caching expert so I get a little lost in the details, but I know this relates to a longstanding issue where the max-age of a block is not bubbled up to page level cache controls. This means that updates may not get to anonymous users or at least that the caches are not invalidated when they should be. https://www.drupal.org/project/drupal/issues/2352009

I believe a better approach is for us to cache this block per route. If something changes on a route (like when a node is updated) then the cached version of that block is updated. So we can replace the `getCacheMaxAge` method with:

```
  use Drupal\Core\Cache\Cache;
  
  
  /**
   * {@inheritdoc}
   */
  public function getCacheContexts() {
    return Cache::mergeContexts(parent::getCacheContexts(), ['route']);
  }
```

Please take a look at this possible solution and let me know if you think it will work.